recirq/kpz/experiment.py

```diff
@@ -0,0 +1,487 @@
+
+    def _generate_initial_states(self) -> np.ndarray:
+        """Generate the initial bitstrings."""
+
+        mu = self.mu
+        n = self.num_qubits
+        nTrials = self.num_init_states
+        if mu == np.inf:
+            bitstrs = np.array(
+                [np.append(np.ones(n // 2, dtype=int), np.zeros(n // 2, dtype=int))]
+            )
+        elif mu < np.inf:
+            p = np.exp(mu) / (np.exp(mu) + np.exp(-mu))
+            bitstrs_L = rng.choice(2, p=[1 - p, p], size=(nTrials, n // 2))
+            bitstrs_R = rng.choice(2, p=[p, 1 - p], size=(nTrials, n // 2))
+            bitstrs = np.append(bitstrs_L, bitstrs_R, axis=1)
+        return bitstrs
+
+    def _generate_cycle(self) -> cirq.Circuit:
+        """Generate the Cirq circuit for one cycle."""
+
+        n = self.num_qubits
+        theta = self.theta
+        phi = self.phi
+        qubits = cirq.LineQubit.range(n)
+
+        qc = cirq.Circuit(
+            cirq.FSimGate(theta, phi)(*qubits[q : q + 2]) for q in range(0, n - 1, 2)
+        )
+        qc += cirq.Circuit(
+            cirq.FSimGate(theta, phi)(*qubits[q : q + 2]) for q in range(1, n - 1, 2)
+        )
+
+        return qc
+
+    def _generate_circuit(self, locs: Iterator[int]) -> cirq.Circuit:
+        """Generate a single circuit.
+
+        Args:
+            locs: Locations of 1 in the initial bitstring.
+
+        Returns:
+            A Cirq circuit that implements the Floquet dynamics for the specified initial state.
+        """
+
+        n = self.num_qubits
+        num_cycles = self.num_cycles
+        qubits = cirq.LineQubit.range(n)
+
+        qc = cirq.Circuit(cirq.X(qubits[q]) for q in locs)
+        if num_cycles > 0:
+            cycle = self._generate_cycle()
+            qc += cycle * num_cycles
+        qc.append(cirq.measure(*qubits, key="m"))
+        return qc
+
+    def _generate_all_circuits(self) -> List[cirq.Circuit]:
+        """Generate Cirq circuits for all of the initial states."""
+
+        locs_all = [np.nonzero(bitstr)[0] for bitstr in self.initial_states]
+        return [self._generate_circuit(locs) for locs in locs_all]
+
+    def run_experiment(self, sampler: cirq.Sampler, reps: int) -> KPZExperimentResults:
+        """Run the experiment using the provided Cirq sampler.
+
+        Args:
+            sampler: The cirq sampler to use for the simulation.
+            reps: The number of bitstrings to sample per initial state.
+
+        Returns:
+            A KPZExperimentResults object containing the measured bitstrings and histogram of
+            transferred magnetization, as well as the extracted moments.
+
+        """
+
+        result = [
+            sampler.run(circuit, repetitions=reps)
+            for circuit in tqdm(self.circuits, total=self.num_init_states)
+        ]
+        return KPZExperimentResults(result, self.initial_states)
+
+    def run_experiment_amplitudes(
+        self, sampler: cirq.Sampler
```

Comment:
Does this actually need a cirq.SimulatesAmplitudes.  For instance, a hardward sampler won't work here, right?

https://quantumai.google/reference/python/cirq/SimulatesAmplitudes